File: poi-api/internal/middleware/ratelimit.go
package middleware

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
)

// rateLimitResponse is the decoded body of an auth-api rate-limit check call.
type rateLimitResponse struct {
	Allowed      bool   `json:"allowed"`
	Remaining    int    `json:"remaining"`
	Limit        int    `json:"limit"`
	ResetsInSecs int64  `json:"resets_in_secs"`
	Error        string `json:"error"`
}

// rateLimitClient is shared across all RateLimit middleware instances to reuse TCP connections.
var rateLimitClient = &http.Client{Timeout: 5 * time.Second}

// Passthrough builds a no-op middleware that forwards every request
// unconditionally (used when AUTH_DISABLED=true). It returns a gin.HandlerFunc
// that always calls Next.
func Passthrough() gin.HandlerFunc {
	return func(c *gin.Context) { c.Next() }
}

// RateLimit validates X-API-Key against the auth-api token bucket, failing
// closed with 503 if auth-api is unreachable. authAPIURL is the base URL of
// the auth-api service, internalSecret is the shared secret used for
// service-to-service HMAC auth, cost is the token cost charged for each
// request, and exempt lists paths that bypass the check entirely. It returns
// a gin.HandlerFunc enforcing the rate limit.
func RateLimit(authAPIURL, internalSecret string, cost int, exempt ...string) gin.HandlerFunc {
	exemptSet := make(map[string]struct{}, len(exempt))
	for _, p := range exempt {
		exemptSet[p] = struct{}{}
	}
	client := rateLimitClient

	return func(c *gin.Context) {
		if _, ok := exemptSet[c.FullPath()]; ok {
			c.Next()
			return
		}

		if h := c.GetHeader("X-Internal-Auth"); h != "" && validInternalAuth(h, internalSecret) {
			c.Next()
			return
		}

		apiKey := c.GetHeader("X-API-Key")
		if apiKey == "" {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "X-API-Key header required"})
			return
		}

		rlResp, err := checkRateLimit(c.Request.Context(), client, authAPIURL, internalSecret, apiKey, cost)
		if err != nil {
			c.AbortWithStatusJSON(http.StatusServiceUnavailable, gin.H{"error": "rate-limit service unavailable"})
			return
		}

		if !rlResp.Allowed {
			if rlResp.Error == "invalid api key" {
				c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "invalid api key"})
				return
			}
			c.Header("X-RateLimit-Remaining", "0")
			c.Header("X-RateLimit-Reset", strconv.FormatInt(
				time.Now().Add(time.Duration(rlResp.ResetsInSecs)*time.Second).Unix(), 10,
			))
			c.Header("Retry-After", strconv.FormatInt(rlResp.ResetsInSecs, 10))
			c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
				"error":          "rate limit exceeded",
				"resets_in_secs": rlResp.ResetsInSecs,
			})
			return
		}

		c.Header("X-RateLimit-Limit", strconv.Itoa(rlResp.Limit))
		c.Header("X-RateLimit-Remaining", strconv.Itoa(rlResp.Remaining))
		c.Header("X-RateLimit-Reset", strconv.FormatInt(
			time.Now().Add(time.Duration(rlResp.ResetsInSecs)*time.Second).Unix(), 10,
		))
		c.Next()
	}
}

// validInternalAuth checks that the HMAC signature matches and the timestamp
// is within ±30 s. header is the value of the X-Internal-Auth header and
// secret is the shared secret used to compute the HMAC. It returns true if
// the header is a valid, fresh signature.
func validInternalAuth(header, secret string) bool {
	parts := strings.SplitN(header, ".", 2)
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return false
	}
	ts, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil {
		return false
	}
	diff := time.Now().Unix() - ts
	if diff > 30 || diff < -30 {
		return false
	}
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(parts[0]))
	expected := hex.EncodeToString(mac.Sum(nil))
	return hmac.Equal([]byte(expected), []byte(parts[1]))
}

// buildInternalAuth builds the "<unix-ts>.<hmac-sha256(secret, ts)>" token
// for service-to-service auth, using secret as the shared HMAC key. It
// returns the signed internal auth token.
func buildInternalAuth(secret string) string {
	ts := strconv.FormatInt(time.Now().Unix(), 10)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(ts))
	return ts + "." + hex.EncodeToString(mac.Sum(nil))
}

// checkRateLimit calls the auth-api internal endpoint to check and consume
// rate-limit tokens. ctx is the request context propagated to the outbound
// call, client is the HTTP client used to reach auth-api, authAPIURL is the
// base URL of the auth-api service, secret is the shared secret for the
// internal auth header, apiKey is the API key being checked, and cost is the
// token cost to charge. It returns the parsed rate-limit response, or an
// error if the call fails.
func checkRateLimit(ctx context.Context, client *http.Client, authAPIURL, secret, apiKey string, cost int) (*rateLimitResponse, error) {
	payload, _ := json.Marshal(map[string]interface{}{
		"api_key": apiKey,
		"cost":    cost,
	})

	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		fmt.Sprintf("%s/internal/check-rate-limit", authAPIURL),
		bytes.NewReader(payload),
	)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Internal-Auth", buildInternalAuth(secret))

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var rl rateLimitResponse
	if err := json.Unmarshal(body, &rl); err != nil {
		return nil, err
	}
	return &rl, nil
}

File: itinerary-api/app/middleware/ratelimit.py
"""Rate-limit middleware: delegates token deduction to auth-api."""

from __future__ import annotations

import hashlib
import hmac
import time

import httpx
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import JSONResponse, Response
from starlette.types import ASGIApp

EXEMPT_PATHS = {"/health"}


def _build_internal_auth(secret: str) -> str:
    """Build a signed X-Internal-Auth header value.

    Args:
        secret: shared secret used to sign the timestamp.

    Returns:
        Header value formatted as '<ts>.<hmac-sha256(secret, ts)>'.
    """
    ts = str(int(time.time()))
    sig = hmac.new(secret.encode(), ts.encode(), hashlib.sha256).hexdigest()
    return f"{ts}.{sig}"


class RateLimitMiddleware(BaseHTTPMiddleware):
    """Calls auth-api /internal/check-rate-limit before each protected request."""

    def __init__(self, app: ASGIApp, auth_api_url: str, internal_secret: str, cost: int) -> None:
        """Configure the middleware and its HTTP client.

        Args:
            app: wrapped ASGI application.
            auth_api_url: base URL of the auth-api service.
            internal_secret: shared secret for internal auth headers.
            cost: token cost charged per request.
        """
        super().__init__(app)
        self._auth_api_url = auth_api_url.rstrip("/")
        self._internal_secret = internal_secret
        self._cost = cost
        self._client = httpx.AsyncClient(timeout=5.0)

    @staticmethod
    def _validate_internal_auth(header: str, secret: str) -> bool:
        """Check whether an X-Internal-Auth header is valid and recent.

        Args:
            header: raw header value formatted as '<ts>.<sig>'.
            secret: shared secret used to verify the signature.

        Returns:
            True if the header is valid and within 30 seconds of now.
        """
        try:
            ts_str, sig = header.split(".", 1)
            ts = int(ts_str)
        except (ValueError, AttributeError):
            return False
        if abs(int(time.time()) - ts) > 30:
            return False
        expected = hmac.new(secret.encode(), ts_str.encode(), hashlib.sha256).hexdigest()
        return hmac.compare_digest(expected, sig)

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        """Enforce rate limiting via auth-api before letting the request through.

        Args:
            request: incoming HTTP request.
            call_next: handler that continues the middleware chain.

        Returns:
            The downstream response, or an error response if blocked.
        """
        if request.url.path in EXEMPT_PATHS:
            return await call_next(request)

        internal_auth = request.headers.get("X-Internal-Auth", "")
        if internal_auth and self._validate_internal_auth(internal_auth, self._internal_secret):
            return await call_next(request)

        api_key = request.headers.get("X-API-Key")
        if not api_key:
            return JSONResponse({"error": "X-API-Key header required"}, status_code=401)

        try:
            resp = await self._client.post(
                f"{self._auth_api_url}/internal/check-rate-limit",
                json={"api_key": api_key, "cost": self._cost},
                headers={"X-Internal-Auth": _build_internal_auth(self._internal_secret)},
            )
            data = resp.json()
        except (httpx.RequestError, httpx.HTTPStatusError, ValueError):
            return JSONResponse({"error": "rate-limit check failed"}, status_code=503)

        if not data.get("allowed"):
            if data.get("error") == "invalid api key":
                return JSONResponse({"error": "invalid api key"}, status_code=401)
            resets_in = data.get("resets_in_secs", 0)
            return JSONResponse(
                {"error": "rate limit exceeded", "resets_in_secs": resets_in},
                status_code=429,
                headers={"Retry-After": str(resets_in)},
            )

        response = await call_next(request)
        response.headers["X-RateLimit-Limit"] = str(data.get("limit", 0))
        response.headers["X-RateLimit-Remaining"] = str(data.get("remaining", 0))
        return response
